### DISCUSSION TITLE
why don't we turn this "f"Pred: {target_image_pred_label} ..." into CPU?

### DISCUSSION THREAD

**Discussion by garylau1** [mrdbourke/pytorch-deep-learning#1005] [Q&A] [answered]
![image](https://github.com/mrdbourke/pytorch-deep-learning/assets/116154053/ce4ccea3-69a8-4502-84df-b4d502edb520)

 plt.imshow(target_image.squeeze().permute(1, 2, 0)) # make sure it's the right size for matplotlib
    if class_names:
        title = f"Pred: {class_names[target_image_pred_label.cpu()]} | Prob: {target_image_pred_probs.max().cpu():.3f}"
    else: 
        title = f"Pred: {target_image_pred_label} | Prob: {target_image_pred_probs.max().cpu():.3f}"
    plt.title(title)
    plt.axis(False);
https://github.com/mrdbourke/pytorch-deep-learning/blob/main/04_pytorch_custom_datasets.ipynb

why don't we turn this "f"Pred: {target_image_pred_label} ..." into CPU as well (?)
Could anyone give me answer as I get confused of it ?
thanks

**Comment by ghost** [comment, marked as answer]
### Example Code:

Here's an improved version of your snippet with explanations:

```python
import matplotlib.pyplot as plt

# Assume `target_image` is a tensor of shape (C, H, W)
plt.imshow(target_image.squeeze().permute(1, 2, 0))  # Make sure the image is in the right format for Matplotlib

# Convert `target_image_pred_label` to CPU if it's a tensor
if isinstance(target_image_pred_label, torch.Tensor):
    target_image_pred_label = target_image_pred_label.cpu().item()

# Format the title based on whether class names are available
if class_names:
    title = f"Pred: {class_names[target_image_pred_label]} | Prob: {target_image_pred_probs.max().cpu():.3f}"
else:
    title = f"Pred: {target_image_pred_label} | Prob: {target_image_pred_probs.max().cpu():.3f}"

plt.title(title)
plt.axis('off')  # Turn off axis
plt.show()
```

### Summary:
- **`target_image_pred_label`**: If this is an integer or scalar value, no need to move it to the CPU. If it were a tensor, you would use `.cpu().item()` to get a Python scalar.
- **`target_image_pred_probs.max().cpu()`**: This is a tensor operation, so you need to move it to the CPU for plotting.

In summary, the key is to understand whether the data is a tensor or a scalar and whether it needs to be moved to the CPU for plotting or other operations.
